File: python/src/problems201_300/delete_node_in_a_linked_list.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def delete_node(self, node: SinglyListNode | None) -> None:
        pass


class Default(Problem):
    @override
    def delete_node(self, node: SinglyListNode | None) -> None:
        if not node:
            return
        nxt: SinglyListNode = node.next
        node.val = nxt.val
        node.next = nxt.next


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems201_300/delete-node-in-a-linked-list.js
class Problem {
  deleteNode(node) {
    throw new Error(`Not implemented (${node}).`);
  }
}

class Default extends Problem {
  deleteNode(node) {
    if (!node) {
      return;
    }
    const next = node.next;
    node.val = next.val;
    node.next = next.next;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
